lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1613,30 +1624,72 @@ class HSAQueue final : public KalmarQueue
             printAsyncOps(std::cerr);
         }
 
+        std::lock_guard<std::recursive_mutex> lg(qmutex);
+
+#if !defined(NDEBUG)
+        bool foundFirstValidOp = false;
+#endif
+        int first_nullptr = -1;
+        int lastWaitOp = asyncOps.size();
+        if (drainingQueue_) {
+            lastWaitOp = (lastWaitOp * QUEUE_FLUSHING_FRAC);
+        }
 
-        {
-            std::lock_guard<std::recursive_mutex> lg(qmutex);
-            bool foundFirstValidOp = false;
-            for (int i = asyncOps.size()-1; i >= 0;  i--) {
-                if (asyncOps[i] != nullptr) {
-                    auto asyncOp = asyncOps[i];
-                    if (!foundFirstValidOp) {
-                        hsa_signal_t sig =  *(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()));
-                        assert(sig.handle != 0);
-                        foundFirstValidOp = true;
-                    }
-                    // wait on valid futures only
-                    std::shared_future<void>* future = asyncOp->getFuture();
-                    if (future && future->valid()) {
-                        future->wait();
-                    }
+        for (int i = lastWaitOp-1; i >= 0;  i--) {
```

Comment:
comment for this loop is necessary I think. the purpose is to find the first op which has become `nullptr`, and trigger the cleaning process in the next block?